Create DevTools Browser Extension Scaffold

noi5e opened the pull request:
- Creates a scaffold for React Components-style panel within DevTools for Chrome, Firefox, and Edge.

<img width="676" alt="Screen Shot 2022-06-10 at 2 57 45 PM" src="https://user-images.githubusercontent.com/4361605/173160730-cb91c4b0-180e-45d0-a6aa-b6aa92a93b62.png">

Currently this is completely unintegrated with Lexical's codebase and build flow.

To run locally:
1. cd `lexical/packages/lexical-devtools`
2. `npm run build`
3. add to Firefox/Chrome/Edge by loading the `build` folder or `manifest.json` as a temporary extension.
  - [example Firefox instructions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Your_first_WebExtension#trying_it_out)

---
Developer Tools Web Extension Planning Issue: #2127

fantactuka commented:
Could you record a video how to run it from chrome then in future we can use it as a reference for anyone who needs help with it?

thegreatercurve commented on packages/lexical-devtools/tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "lib": [
+      "dom",
+      "dom.iterable",
+      "esnext"
+    ],
+    "allowJs": true,
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "strict": true,
+    "forceConsistentCasingInFileNames": true,
+    "noFallthroughCasesInSwitch": true,
+    "module": "esnext",
+    "moduleResolution": "node",
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "noEmit": true,
+    "jsx": "react-jsx"
+  },
```
This can probably just inherit from the base `tsconfig` to avoid having different TS configs per package.

```suggestion
  "compilerOptions": {
    "strict": true,
  },
  "extends": "../tsconfig.json" 
```

noi5e replied:
> This can probably just inherit from the base tsconfig to avoid having different TS configs per package.

Sounds good, will commit

thegreatercurve commented on packages/lexical-devtools/public/devtools.js:
```diff
@@ -0,0 +1,4 @@
+'use strict';
+
+// eslint-disable-next-line no-undef
+chrome.devtools.panels.create('Lexical', 'logo192.png', 'index.html', () => {});
```
To remove the lint warning: 
```suggestion
chrome.devtools.panels.create('Lexical', 'logo192.png', 'index.html', () => {
    return;
});
```

noi5e replied:
> To remove the lint warning:

@thegreatercurve Ah yeah, ESLint was actually signalling `chrome` as undefined here. It seems like `chrome` is defined in Firefox, Chrome, and Edge. Maybe `let chrome = chrome` would be a way to get around the linter without disabling it? Open to suggestions here.

fantactuka replied:
Maybe similar to this: https://github.com/facebook/lexical/blob/c8b7add27d1435972a586a1c89130ba776748914/.eslintrc.js#L14

thegreatercurve reviewed (approved):
Looks good! Thanks for working on this.

noi5e commented:
Hmm, the CI tests have been failing:
<img width="838" alt="Screen Shot 2022-06-12 at 12 17 30 PM" src="https://user-images.githubusercontent.com/4361605/173249718-c32515b9-777f-4776-81dd-3531bfdf356e.png">

Due to installing dependencies:
<img width="1041" alt="Screen Shot 2022-06-12 at 12 07 52 PM" src="https://user-images.githubusercontent.com/4361605/173249735-ce377926-1f6d-4e57-b9b5-34954441ba6f.png">

I made this PR by creating a web extension scaffold **in a separate repo** [here](https://github.com/noi5e/lexical-dev-tools). Then I copied-and-pasted what I had into a new folder under `packages`. 

Maybe something about the config/dependencies from **that repo** is not playing nice with existing Lexical config.

I'm noticing that other Lexical `packages` have their own individual `package.json`. But it seems like they don't have a `package-lock.json` like this PR has.

This is kind of a shot in the dark, but I wonder if deleting the `package-lock.json` here will help this pass. Will play around with it.

noi5e commented:
> Could you record a video how to run it from chrome then in future we can use it as a reference for anyone who needs help with it? I'd probably even put it into readme/contribution section

@fantactuka In the next PR, I hope to fully integrate the package with Lexical's build process, with Vite etc. The goal in the next few days or so is to have the browser extension build and run automatically without any tinkering required. So hopefully having to run `npm run build` manually is just a temporary state of affairs 🤞.

I just included those instructions so maintainers could see what the scaffold looked like currently... But I made a video anyway!

https://user-images.githubusercontent.com/4361605/173254833-48af086c-abf5-40c0-b849-e7b09b283956.mov

> 1. cd packages/lexical-devtools
> 2. npm run build
> 3. add to Firefox/Chrome/Edge by loading the build folder or manifest.json as a temporary extension.

noi5e commented:
<img width="840" alt="Screen Shot 2022-06-12 at 4 20 51 PM" src="https://user-images.githubusercontent.com/4361605/173257750-3de459f0-71f2-47fb-b963-8a2f768ebdec.png">

Since I'm a first-time contributor, my commits don't trigger GitHub Actions... @thegreatercurve @fantactuka Can you run the workflow for me when you get a chance?

I deleted `package-lock.json` to see if that would help this pass the CI, let's see. 🙏 

I also added build instructions in `README.md`.

And changed `tsconfig.json` to inherit from base. I believe I can just delete `tsconfig.json`, but for some reason when I do that, the extension won't `build` properly. In the next PR, I'm going to try to build the package with Vite instead of Create React App, so I'll try again there.

noi5e commented:
Checks are passing now 🥳 :
<img width="841" alt="Screen Shot 2022-06-12 at 5 33 07 PM" src="https://user-images.githubusercontent.com/4361605/173260262-356ad105-24de-4113-9926-ff9a5b769696.png">

thegreatercurve commented:
@noi5e I'll merge this and you can create experiment with Vite in another separate PR.
